crates/sui-package-resolver/src/lib.rs

```diff
@@ -1840,6 +1855,42 @@ mod tests {
         assert!(matches!(err, Error::TooManyTypeParams(1, 2)));
     }
 
+    #[tokio::test]
+    async fn test_err_too_many_type_nodes() {
+        use Ability as A;
+        use AbilitySet as S;
+
+        let (_, cache) = package_cache([
+            (1, build_package("sui"), sui_types()),
+            (1, build_package("d0"), d0_types()),
+        ]);
+
+        let resolver = Resolver::new_with_limits(
+            cache,
+            Limits {
+                max_type_argument_width: 100,
+                max_type_argument_depth: 100,
+                max_type_nodes: 2,
+                max_move_value_depth: 100,
+            },
+        );
+
+        // This request is OK, because one of O's type parameters is phantom, so we can avoid
+        // loading its definition.
+        let a1 = resolver
+            .abilities(type_("0xd0::m::O<0xd0::m::S, 0xd0::m::Q>"))
+            .await
+            .unwrap();
+        assert_eq!(a1, S::EMPTY | A::Key | A::Store);
+
+        // But this request will hit the limit
+        let err = resolver
+            .abilities(type_("0xd0::m::T<0xd0::m::P, 0xd0::m::Q>"))
```

Comment:
question, not blocking: this fails because we need to look up T, P, and Q?

Reply:
Correct -- and the previous request does not because one of `O`'s type parameters is `phantom` so we don't need to look it up.